File: src/main/java/ch/softappeal/yass/serialize/Serializer.java
package ch.softappeal.yass.serialize;

import ch.softappeal.yass.util.Nullable;

/**
 * Writes and reads values.
 */
public interface Serializer {

    @Nullable Object read(Reader reader) throws Exception;

    void write(@Nullable Object value, Writer writer) throws Exception;

}

File: kotlin/yass/main/serialize/Serializer.kt
package ch.softappeal.yass.serialize

import java.io.ObjectInputStream
import java.io.ObjectOutputStream

interface Serializer {
    @Throws(Exception::class)
    fun read(reader: Reader): Any?

    @Throws(Exception::class)
    fun write(writer: Writer, value: Any?)
}

val JavaSerializer = object : Serializer {
    override fun read(reader: Reader) =
        ObjectInputStream(reader.stream()).use(ObjectInputStream::readObject)

    override fun write(writer: Writer, value: Any?) =
        ObjectOutputStream(writer.stream()).use { it.writeObject(value) }
}
